File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char (isUpper, isLetter, isSpace)
import Data.List (dropWhileEnd)

responseFor :: String -> String
responseFor xs 
	| silence  = "Fine. Be that way!"
	| shout = "Whoa, chill out!"
	| question = "Sure."
	| otherwise = "Whatever."
	where 
		silence = all isSpace xs
		shout = every isUpper (filter isLetter xs)
		question = last (dropWhileEnd isSpace xs) == '?'

every :: (a -> Bool) -> [a] -> Bool
every _ [] = False
every pred list = all pred list

File: clojure/bob/src/bob.clj
(ns bob)

(defn every [fn msg]
	( and 
		(every? fn msg) 
		(not (empty? msg))
	)
)

(defn response-for [msg] 
	(cond 
		(clojure.string/blank? msg) "Fine. Be that way!"
		(every 
			#(Character/isUpperCase %) 
			(filter #(Character/isLetter %) msg)
		) "Whoa, chill out!"
		(= \? (last (clojure.string/trim msg))) "Sure."
		:else "Whatever."
	)
)
